// kazagumo-apple/dist/plugin.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.KazagumoPlugin = void 0;
const kazagumo_1 = require("kazagumo");
const axios = require("axios");
const credentials_1 = require("./credentials");
const REGEX = /(?:https:\/\/music\.apple\.com\/)(?:.+)?(artist|album|music-video|playlist)\/([\w\-\.]+(\/)+[\w\-\.]+|[^&]+)\/([\w\-\.]+(\/)+[\w\-\.]+|[^&]+)/;
const REGEX_SONG_ONLY = /(?:https:\/\/music\.apple\.com\/)(?:.+)?(artist|album|music-video|playlist)\/([\w\-\.]+(\/)+[\w\-\.]+|[^&]+)\/([\w\-\.]+(\/)+[\w\-\.]+|[^&]+)(\?|\&)([^=]+)\=([\w\-\.]+(\/)+[\w\-\.]+|[^&]+)/;
class KazagumoPlugin extends kazagumo_1.KazagumoPlugin {
    constructor(appleOptions) {
        var _a, _b, _c, _d, _e, _f;
        super();
        this.methods = {
            artist: this.getArtist.bind(this),
            album: this.getAlbum.bind(this),
            playlist: this.getPlaylist.bind(this),
            track: this.getTrack.bind(this),
        };
        this.options = appleOptions;
        this.kazagumo = null;
        this._search = null;
        this.countryCode = ((_a = this.options) === null || _a === void 0 ? void 0 : _a.countryCode)
            ? (_b = this.options) === null || _b === void 0 ? void 0 : _b.countryCode
            : "us";
        this.imageHeight = ((_c = this.options) === null || _c === void 0 ? void 0 : _c.imageHeight)
            ? (_d = this.options) === null || _d === void 0 ? void 0 : _d.imageHeight
            : 900;
        this.imageWidth = ((_e = this.options) === null || _e === void 0 ? void 0 : _e.imageWidth) ? (_f = this.options) === null || _f === void 0 ? void 0 : _f.imageWidth : 600;
        this.baseURL = "https://api.music.apple.com/v1/";
        this.fetchURL = `https://amp-api.music.apple.com/v1/catalog/${this.countryCode}`;
        this.credentials = {
            Authorization: `Bearer ${credentials_1.credentials.APPLE_TOKEN}`,
            origin: "https://music.apple.com",
        };
    }
    load(kazagumo) {
        this.kazagumo = kazagumo;
        this._search = kazagumo.search.bind(kazagumo);
        kazagumo.search = this.search.bind(this);
    }
    getData(params) {
        return __awaiter(this, void 0, void 0, function* () {
            let req = yield axios.get(`${this.fetchURL}${params}`, {
                headers: this.credentials,
            });
            return req.data;
        });
    }
    search(query, options) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            let type;
            let id;
            let is_track = false;
            if (!this.kazagumo || !this._search)
                throw new kazagumo_1.KazagumoError(1, "kazagumo-apple is not loaded yet.");
            if (!query)
                throw new kazagumo_1.KazagumoError(3, "Query is required");
            if (!REGEX_SONG_ONLY.exec(query) || REGEX_SONG_ONLY.exec(query) == null) {
                const extract = REGEX.exec(query) || [];
                id = extract[4];
                type = extract[1];
            }
            else {
                const extract = REGEX_SONG_ONLY.exec(query) || [];
                id = extract[8];
                type = extract[1];
                is_track = true;
            }
            // const [, type, id] = REGEX.exec(query) || [];
            const isUrl = /^https?:\/\//.test(query);
            if (type in this.methods) {
                try {
                    let _function = this.methods[type];
                    if (is_track)
                        _function = this.methods.track;
                    const result = yield _function(id, options === null || options === void 0 ? void 0 : options.requester);
                    const loadType = is_track ? "TRACK" : "PLAYLIST";
                    const playlistName = (_a = result.name) !== null && _a !== void 0 ? _a : undefined;
                    const tracks = result.tracks.filter(this.filterNullOrUndefined);
                    return this.buildSearch(playlistName, tracks, loadType);
                }
                catch (e) {
                    return this.buildSearch(undefined, [], "SEARCH");
                }
            }
            else if ((options === null || options === void 0 ? void 0 : options.engine) === "apple" && !isUrl) {
                const result = yield this.searchTrack(query, options === null || options === void 0 ? void 0 : options.requester);
                return this.buildSearch(undefined, result.tracks, "SEARCH");
            }
            return this._search(query, options);
        });
    }
    buildSearch(playlistName, tracks = [], type) {
        return {
            playlistName,
            tracks,
            type: type !== null && type !== void 0 ? type : "TRACK",
        };
    }
    searchTrack(query, requester) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const res = yield this.getData(`/search?types=songs&term=${query
                    .replace(/ /g, "+")
                    .toLocaleLowerCase()}`).catch((e) => {
                    throw new Error(e);
                });
                return {
                    tracks: res.results.songs.data.map((track) => this.buildKazagumoTrack(track, requester)),
                };
            }
            catch (e) {
                throw new Error(e);
            }
        });
    }
    getTrack(id, requester) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const track = yield this.getData(`/songs/${id}`).catch((e) => {
                    throw new Error(e);
                });
                return { tracks: [this.buildKazagumoTrack(track.data[0], requester)] };
            }
            catch (e) {
                throw new Error(e);
            }
        });
    }
    getArtist(id, requester) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const track = yield this.getData(`/artists/${id}/view/top-songs`).catch((e) => {
                    throw new Error(e);
                });
                return { tracks: [this.buildKazagumoTrack(track.data[0], requester)] };
            }
            catch (e) {
                throw new Error(e);
            }
        });
    }
    getAlbum(id, requester) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const album = yield this.getData(`/albums/${id}`).catch((e) => {
                    throw new Error(e);
                });
                const tracks = album.data[0].relationships.tracks.data
                    .filter(this.filterNullOrUndefined)
                    .map((track) => this.buildKazagumoTrack(track, requester));
                return { tracks, name: album.data[0].attributes.name };
            }
            catch (e) {
                throw new Error(e);
            }
        });
    }
    getPlaylist(id, requester) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const playlist = yield this.getData(`/playlists/${id}`).catch((e) => {
                    throw new Error(e);
                });
                const tracks = playlist.data[0].relationships.tracks.data
                    .filter(this.filterNullOrUndefined)
                    .map((track) => this.buildKazagumoTrack(track, requester));
                return { tracks, name: playlist.data[0].attributes.name };
            }
            catch (e) {
                throw new Error(e);
            }
        });
    }
    filterNullOrUndefined(obj) {
        return obj !== undefined && obj !== null;
    }
    buildKazagumoTrack(appleTrack, requester) {
        const artworkURL = new String(appleTrack.attributes.artwork.url)
            .replace("{w}", String(this.imageWidth))
            .replace("{h}", String(this.imageHeight));
        return new kazagumo_1.KazagumoTrack({
            encoded: "",
            info: {
                sourceName: "apple",
                identifier: appleTrack.id,
                isSeekable: true,
                author: appleTrack.attributes.artistName
                    ? appleTrack.attributes.artistName
                    : "Unknown",
                length: appleTrack.attributes.durationInMillis,
                isStream: false,
                position: 0,
                title: appleTrack.attributes.name,
                uri: appleTrack.attributes.url || "",
                artworkUrl: artworkURL ? artworkURL : "",
            },
            pluginInfo: null,
        }, requester);
    }
}
exports.KazagumoPlugin = KazagumoPlugin;
